File: 백준/Gold/14503. 로봇 청소기/로봇 청소기.java
import java.io.*;
import java.util.*;

public class Main {
    // 방향 벡터 (북, 동, 남, 서)
    private static final int[] dx = {-1, 0, 1, 0};
    private static final int[] dy = {0, 1, 0, -1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        int r = Integer.parseInt(st.nextToken());
        int c = Integer.parseInt(st.nextToken());
        int d = Integer.parseInt(st.nextToken());

        int[][] room = new int[N][M];
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < M; j++) {
                room[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        System.out.println(cleanRoom(N, M, r, c, d, room));
    }

    private static int cleanRoom(int N, int M, int r, int c, int d, int[][] room) {
        int cleanedCount = 0;

        while (true) {
            // 1. 현재 칸이 청소되지 않았다면 청소
            if (room[r][c] == 0) {
                room[r][c] = 2; // 청소 완료 표시
                cleanedCount++;
            }

            // 2. 현재 칸의 주변 4칸 탐색
            boolean found = false;
            for (int i = 0; i < 4; i++) {
                d = (d + 3) % 4; // 반시계 방향으로 90도 회전
                int nx = r + dx[d];
                int ny = c + dy[d];

                if (nx >= 0 && nx < N && ny >= 0 && ny < M && room[nx][ny] == 0) {
                    r = nx;
                    c = ny;
                    found = true;
                    break;
                }
            }

            // 3. 청소되지 않은 빈 칸이 없는 경우
            if (!found) {
                int backDirection = (d + 2) % 4;
                int bx = r + dx[backDirection];
                int by = c + dy[backDirection];

                if (bx >= 0 && bx < N && by >= 0 && by < M && room[bx][by] != 1) {
                    r = bx;
                    c = by;
                } else {
                    break; // 뒤가 벽인 경우 작동 중지
                }
            }
        }

        return cleanedCount;
    }
}

File: 백준/Gold/14503. 로봇 청소기/로봇 청소기.py
def robot_vacuum_cleaner(N, M, r, c, d, room):
    # 방향 벡터 (북, 동, 남, 서)
    directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]

    # 청소한 칸 개수
    cleaned_count = 0

    # 현재 위치 및 방향
    x, y, direction = r, c, d

    while True:
        # 1. 현재 칸이 청소되지 않았다면 청소
        if room[x][y] == 0:
            room[x][y] = 2  # 청소 완료 표시
            cleaned_count += 1

        # 2. 현재 칸의 주변 4칸 탐색
        found = False
        for _ in range(4):
            # 반시계 방향으로 90도 회전
            direction = (direction + 3) % 4
            nx, ny = x + directions[direction][0], y + directions[direction][1]

            # 청소되지 않은 빈 칸이 있는 경우 전진
            if room[nx][ny] == 0:
                x, y = nx, ny
                found = True
                break

        # 3. 청소되지 않은 빈 칸이 없는 경우
        if not found:
            # 뒤로 이동할 수 있으면 이동
            back_direction = (direction + 2) % 4
            bx, by = x + directions[back_direction][0], y + directions[back_direction][1]
            if room[bx][by] != 1:
                x, y = bx, by
            else:
                # 뒤가 벽인 경우 작동 중지
                break

    return cleaned_count

# 입력 처리
N, M = map(int, input().split())
r, c, d = map(int, input().split())
room = [list(map(int, input().split())) for _ in range(N)]

# 결과 출력
print(robot_vacuum_cleaner(N, M, r, c, d, room))